### PR TITLE
Add more types for "repl" and "module" modules in Node

### PR DESCRIPTION
**Pull request by devoto13**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
    Added APIs are used by [ts-node](https://github.com/TypeStrong/ts-node), but mostly not mentioned in the documentation pages.
    [ouputStream/inputStream](https://github.com/nodejs/node/blob/a80b1621b034b6bcd920805681db22e1f6c5a282/lib/repl.js#L326)
    [Recoverable](https://github.com/nodejs/node/blob/a80b1621b034b6bcd920805681db22e1f6c5a282/lib/repl.js#L1388)
    ["module" module, Module class](https://github.com/nodejs/node/blob/a80b1621b034b6bcd920805681db22e1f6c5a282/lib/module.js#L55).
    `require.extensions` can be checked by running `require.extensions['.js'].toString()` and they are used by `ts-node` this way.
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by devoto13** [comment]
Any ideas how to fix the build? I don't see how my changes could cause this error. If I run `tsc` locally in `whatwg-stream` package on `master` it fails with the same error without my changes either, but build in CI is green...

File: types/node/index.d.ts [hunk 1]
```diff
@@ -5592,6 +5608,31 @@ declare module "constants" {
     export var ALPN_ENABLED: number;
 }
 
+declare module "module" {
+    class Module implements NodeModule {
+        static runMain(): void;
+        static wrap(code: string): string;
+        static _nodeModulePaths(path: string): string[];
+        static _load(request: string, parent?: NodeModule, isMain?: boolean): any;
```
Comment (Flarna) [inline types/node/index.d.ts]: As far as I know non documented, internal APIs are not added to typings, only public APIs.

Comment (devoto13) [reply types/node/index.d.ts]: Seems you're right. Don't see many private APIs in other typings. I've removed them.

**Comment by devoto13** [comment]
@andy-ms any timeline for fixing `whatwg-stream` in master? I get same errors without my changes, so I don't think they are reason of the build failure.

From Travis logs:

```
Error in whatwg-streams
Command failed: node /home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/index.js --noLint
Error in TypeScript@next: ../../node_modules/dtslint/typescript-installs/next/node_modules/typescript/lib/lib.dom.d.ts(9031,11): error TS2300: Duplicate identifier 'ReadableStream'.
../../node_modules/dtslint/typescript-installs/next/node_modules/typescript/lib/lib.dom.d.ts(9037,13): error TS2300: Duplicate identifier 'ReadableStream'.
index.d.ts(32,15): error TS2300: Duplicate identifier 'ReadableStream'.
whatwg-streams-tests.ts(12,16): error TS2554: Expected 0 arguments, but got 1.
whatwg-streams-tests.ts(13,19): error TS7006: Parameter 'controller' implicitly has an 'any' type.
whatwg-streams-tests.ts(31,16): error TS7006: Parameter 'e' implicitly has an 'any' type.
whatwg-streams-tests.ts(40,12): error TS2554: Expected 0 arguments, but got 1.
whatwg-streams-tests.ts(41,15): error TS7006: Parameter 'controller' implicitly has an 'any' type.
whatwg-streams-tests.ts(79,12): error TS2554: Expected 0 arguments, but got 1.
whatwg-streams-tests.ts(82,15): error TS7006: Parameter 'controller' implicitly has an 'any' type.
whatwg-streams-tests.ts(137,16): error TS2554: Expected 0 arguments, but got 1.
whatwg-streams-tests.ts(144,18): error TS7006: Parameter 'controller' implicitly has an 'any' type.
whatwg-streams-tests.ts(175,16): error TS2554: Expected 0 arguments, but got 1.
whatwg-streams-tests.ts(184,18): error TS7006: Parameter 'controller' implicitly has an 'any' type.
whatwg-streams-tests.ts(240,20): error TS2339: Property 'pipeTo' does not exist on type 'ReadableStream'.
whatwg-streams-tests.ts(242,16): error TS7006: Parameter 'e' implicitly has an 'any' type.
whatwg-streams-tests.ts(291,23): error TS2554: Expected 0 arguments, but got 1.
whatwg-streams-tests.ts(351,27): error TS7031: Binding element 'value' implicitly has an 'any' type.
whatwg-streams-tests.ts(351,34): error TS7031: Binding element 'done' implicitly has an 'any' type.
```

**Review by Flarna — APPROVED** [review]

**Comment by andy-ms** [comment]
@devoto13 Looks like that is blocked by Microsoft/TSJS-lib-generator#222 -- the current `lib.d.ts` uses `var` to declare `ReadableStream`, making it impossible to add any new declarations to the static side. @mhegazy Should we accept a PR for that?

File: types/node/index.d.ts [hunk 2]
```diff
@@ -93,10 +93,17 @@ interface NodeRequireFunction {
 interface NodeRequire extends NodeRequireFunction {
     resolve(id: string): string;
     cache: any;
-    extensions: any;
+    extensions: NodeExtensions;
     main: NodeModule | undefined;
 }
 
+interface NodeExtensions {
+  '.js': (m: NodeModule, filename: string) => any;
```
Comment (andy-ms) [inline types/node/index.d.ts]: These all have the same type of value, shouldn't it just be the string indexer then?

Comment (devoto13) [reply types/node/index.d.ts]: I think it's better to leave them for documentation purposes. From type checker point of view it doesn't matter, but for human reader it should make it more clear what is expected usage.

**Comment by develar** [comment]
It leads to compilation error:

```
    (5615,11): error TS2420: Class 'Module' incorrectly implements interface 'NodeModule'.
      Property 'hot' is missing in type 'Module'.
```

if `@types/webpack-env` also used https://www.npmjs.com/package/@types/webpack-env Not clear for me what and where (this or webpack-env) should be fixed :)

**Comment by devoto13** [comment]
@develar See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19602.

**Comment by Flarna** [comment]
My feeling here is that the problem is in webpack-env as it adds mandatory properties to the global interface exported by Node which are most likely fine in WebPack context but not for node.

**Comment by tko** [comment]
Best I can tell I don't have any webpack related typescript definitions anywhere and yet getting the error about 'hot' property. See https://github.com/DefinitelyTyped/DefinitelyTyped/issues/19623

**Comment by Flarna** [comment]
@tko, @develar: I created already a PR which should fix the issues you see: #19612

**Pull request merged**